### DISCUSSION TITLE
Decision support EspAsyncWebserver vs. PsychicHTTP

### DISCUSSION THREAD

**Discussion by dewenni** [hoeken/PsychicHttp#197] [General]
Since I can imagine that there are several users with this question, I would like to ask it here.
I am currently using what is probably the best maintained version of the EspAsync web server from @mathieucarbou.
In my project I have used SSE so far and have currently switched to Websockets due to stability problems.
This seems to work more stable so far. Nevertheless, I am wondering whether I should make further adjustments to achieve the most stable solution possible.
I am also asking myself whether I should continue to use EspAsyncWebserver+AsyncTCP or perhaps this alternative here.
Nobody can probably give me a clear answer to this question, but I'm finding it difficult to make a technical evaluation.
Therefore my question:
- What are the main differences between EspAsyncWebserver and PsychicHTTP?
- As far as I can see, you can realise the basics (web server, SSE, WS) with both solutions, but what are the particular strengths and weaknesses?
- PsychicHTTP was initially ‘advertised’ as stable and robust - how can this be argued?

In my personal use case, I need a web server whose content is loaded from the PROGMEM.
I think this is generally not a challenge. After connecting a client, however, larger JSON packets have to be sent to the client (4kB + 2kB + 2kB) and then event-triggered possibly many small messages at intervals of approx. 20ms. Several clients can also be connected (max. 2-3). Besides that I need OTA and File Up and Download.

**Comment by mathieucarbou** [comment]
Hello @dewenni !

FYI,  I am also a heavy contributor of Psychic (actually one of its collaborator). I've worked on the middleware support and some API refinement in the v2. You can join our Discord here: https://discord.gg/TAQrTR3f9C

Here are some key differences that for me can be the reason why to choose one or the other.

- Psychic supports HTTPS while ESPAsyncWS is not

- Psychic is based on the ESP-IDF http server API which is supported by Espressif and is reliable

- Psychic is new compared to ESPAsyncWS, so there are not a lot of libraries using it.

- Psychic is better built (although there still a lot of API improvements to do)  and based on more reliable base. Psychic is just a wrapper around the esp-idf api to make it easier to use.

- Current benchmark tests prove that Psychic can handle more concurrent requests and faster then ESPAsyncWS, but ESPAsyncWS is better at WebSockets (this though is if I recall only linked to some esp-idf limitations that could be improved later). 

- Psychic is only ESP32 while ESPAsyncWS 3.x keeps compatibility with ESP8266 and RPI2040, which could be a deciding factor

- Both ESPAsyncWS and Psychic are OSS projects. Psychic v2 is still in development, really stable but there might still be some API change happening. If you choose Psychic, go with v2.

- ESPAsyncWS 3.x has some API limitations and constraints due to being compatible with ESP32, ESP8266 and RPI2040. As I have written recently in a ESPAsyncWS discussion, version 4.x will have its API refined, will be faster and will drop support for ESP8266, RPI40 and ArduinoJson versions below 7. Version 4.x will not be backward compatible with 3.x

- Psychic dev is really slow, because we (both collaborators) are improving it on our spare time, while ESPAsyncWS 3.x is more in "maintenance" mode and won't have anything more except a big lifting in version 4.x.

- ESPAsyncWS is based on AsyncTCP layer (lwip), and the Arduino Core team have plans to bring AsyncTCP under the Espressif umbrella, exactly like they currently have a AsyncUDP component in Arduino Core, and officially improve and support it. So https support will for sure come once this step will be done and everything using AsyncTCP like ESPAsyncWS will then run on the Arduino Core (Espressif) supported AsyncTCP

- ESPAsyncWS being widely spread, it has years of fixes and XP and user base behind it, but it also has unfixable limitations. There are many libs and projects using it, so it will never go away. And for example in some app i have developed, I will never be able to use Psychic because I am using too many key libraries relying on ESPAsyncWS like ESP-DASH, ElegantOTA, MycilaWebSerial, MycilaESPConnect, etc.

- AsyncTCP and ESPAsyncWS are made by me-no-dev, who is working at Espressif in the Arduino Core team. When they'll bring AsyncTCP under their home, we should see improvements in everything linked to this base, so all the projects around based on AsyncTCP either for client side or server side will gain benefits of that.

- Psychic is made by some more random people having contributed to the project over time on our spare time. Which means slow dev, more risk at seeing the project left for some months without commits.

**So based on that, here are my advices:**

- If you really need https support for server, go with Psychic
- If your app needs to rely on some mqtt libs or other libs based on ESPAsyncWS or AsyncTCP, then go with ESPAsyncWS.
- If you have a complete choice, and not the contrains below, go with Psychic
- If your app heavily uses WebSockets, go with ESPAsyncWS (it is my understanding that ESPAsyncWS is very good for WebSockets and beats Psychic but this will chance for sure one day)
- If your app has memory constraints, go with ESPAsyncWS: ESPAsyncWS is using less heap / stack overall: Psychic is spawning workers to handler a lot of concurrent requests which consumes a lot of stack and heap.
- If your app requires to serve a lot of concurrent requests fast, go with Psychic. Look at the benchmark on the [ESPAsyncWS maintained fork](https://github.com/mathieucarbou/ESPAsyncWebServer?tab=readme-ov-file#performance) and this project.
- If your app requires some sort of long term support and  requires to be more resilient to future Arduino Core or ESP-IDF API changes, then go with AsyncTCP and ESPAsyncWS because the user base is so huge that any issues seen during the RC releases are fixed quite fast. If your app can continue to work on the same versions all the time then you do not have this upgrade concern.
